File: Algorithms/C++/Medium/144_Binary_Tree_Preorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> values;

        if (root == NULL) {
            return values;
        }

        stack<TreeNode*> s;
        s.push(root);

        while (!s.empty()) {
            TreeNode* node = s.top();
            s.pop();

            values.push_back(node->val);

            if (node->right != NULL) {
                s.push(node->right);
            }
            if (node->left != NULL) {
                s.push(node->left);
            }
        }

        return values;
    }
};

File: Algorithms/JavaScript/Medium/144_Binary_Tree_Preorder_Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    var values = [];

    if (root === null) {
        return values;
    }

    var stack = [];
    stack.unshift(new Pair(null, root, Type.Node));

    while (stack.length > 0) {
        var top = stack.shift();
        if (top.type === Type.Value) {
            values.push(top.val);
        } else {
            var node = top.node;
            if (node.right !== null) {
                stack.unshift(new Pair(null, node.right, Type.Node));
            }
            if (node.left !== null) {
                stack.unshift(new Pair(null, node.left, Type.Node));
            }
            stack.unshift(new Pair(node.val, null, Type.Value));
        }
    }

    return values;
};

var Type = {
    Value: 0,
    Node: 1
};

function Pair(val, node, type) {
    this.val = val;
    this.node = node;
    this.type = type;
}
